File: auth_service/internal/transport/helpers/validator.go
package helpers

import (
	errs "github.com/NeF2le/anonix/common/errors"
	"unicode"
)

func ValidatePassword(pw string) error {
	if len(pw) > 72 {
		return errs.ErrPasswordTooLong
	}
	if len(pw) < 8 {
		return errs.ErrPasswordTooShort
	}

	var hasLetter, hasDigit bool

	for _, r := range pw {
		if r > unicode.MaxASCII {
			return errs.ErrPasswordNonASCII
		}

		switch {
		case unicode.IsLetter(r):
			hasLetter = true
		case unicode.IsDigit(r):
			hasDigit = true
		}
	}

	if !hasLetter || !hasDigit {
		return errs.ErrPasswordWeak
	}

	return nil
}

File: gateway/static/scripts/helpers/validator.js
function validatePassword(pw) {
    if (pw.length > 72) {
        return "Слишком длинный пароль";
    }
    if (pw.length < 8) {
        return "Пароль должен быть от 8 символов";
    }

    let hasLetter = false;
    let hasDigit = false;

    for (let i = 0; i < pw.length; i++) {
        const char = pw[i];
        const code = char.charCodeAt(0);

        if (code > 127) {
            return "Недопустимые символы в пароле";
        }

        if ((code >= 65 && code <= 90) || (code >= 97 && code <= 122)) {
            hasLetter = true;
        } else if (code >= 48 && code <= 57) {
            hasDigit = true;
        }
    }

    if (!hasLetter || !hasDigit) {
        return "Пароль должен содержать латинские буквы и цифры";
    }

    return "";
}
